File: CombatCore/MatchState.cs
namespace CombatCore;

public enum MatchPhase : byte
{
    Fighting = 0,
    RoundEnd = 1,
}

/// <summary>
/// The complete, authoritative snapshot of one instant of the match. This is
/// the type that:
///   - the server treats as ground truth and broadcasts (compressed - see NetCodec)
///   - the client predicts locally, ahead of server confirmation
///   - the rollback controller snapshots every frame so it can rewind and replay
///   - StateHash.Compute() fingerprints for desync detection
///
/// It is a struct containing only structs (PlayerState is itself a struct), so
/// copying a MatchState is a flat memory copy with no heap allocation and no
/// aliasing surprises - important for a rollback buffer that might keep 10+
/// seconds of history.
/// </summary>
public struct MatchState
{
    public uint FrameNumber;
    public MatchPhase Phase;
    public PlayerState Player1;
    public PlayerState Player2;
    public DeterministicRandom Rng;

    public static MatchState CreateInitial(uint rngSeed)
    {
        return new MatchState
        {
            FrameNumber = 0,
            Phase = MatchPhase.Fighting,
            Player1 = PlayerState.CreateDefault(FixedVector3.FromFloats(-1.5f, 0f, 0f), facingSign: 1),
            Player2 = PlayerState.CreateDefault(FixedVector3.FromFloats(1.5f, 0f, 0f), facingSign: -1),
            Rng = new DeterministicRandom(rngSeed),
        };
    }
}

File: Client/src/sim/MatchState.ts
import { FixedVector3 } from "./FixedVector3.ts";
import { DeterministicRandom } from "./DeterministicRandom.ts";
import { createDefaultPlayerState, clonePlayerState } from "./PlayerState.ts";
import type { PlayerState } from "./PlayerState.ts";

export const MatchPhase = {
  Fighting: 0,
  RoundEnd: 1,
} as const;
export type MatchPhaseValue = (typeof MatchPhase)[keyof typeof MatchPhase];

export interface MatchState {
  frameNumber: number;
  phase: MatchPhaseValue;
  player1: PlayerState;
  player2: PlayerState;
  rng: DeterministicRandom;
}

export function createInitialMatchState(rngSeed: number): MatchState {
  return {
    frameNumber: 0,
    phase: MatchPhase.Fighting,
    player1: createDefaultPlayerState(FixedVector3.fromFloats(-1.5, 0, 0), 1),
    player2: createDefaultPlayerState(FixedVector3.fromFloats(1.5, 0, 0), -1),
    rng: new DeterministicRandom(rngSeed),
  };
}

/** Deep-enough clone for a rollback snapshot buffer - matches C#'s struct-copy semantics. */
export function cloneMatchState(s: MatchState): MatchState {
  return {
    frameNumber: s.frameNumber,
    phase: s.phase,
    player1: clonePlayerState(s.player1),
    player2: clonePlayerState(s.player2),
    rng: s.rng.clone(),
  };
}
